### PR TITLE
Fix false positive on `semicolon_if_nothing_returned`

### LINKED ISSUE #7324: semicolon_if_nothing_returned false positive inside unsafe block
In the following code, I believe it is correct as written, and a semicolon inside the block is not warranted.

```rust
use std::mem::MaybeUninit;
use std::ptr;

fn main() {
    let mut s = MaybeUninit::<String>::uninit();
    let _d = || unsafe { ptr::drop_in_place(s.as_mut_ptr()) };
}
```

```console
$ cargo clippy -- -Dclippy::pedantic

error: consider adding a `;` to the last statement for consistent formatting
 --> src/main.rs:6:26
  |
6 |     let _d = || unsafe { ptr::drop_in_place(s.as_mut_ptr()) };
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `ptr::drop_in_place(s.as_mut_ptr());`
  |
  = note: `-D clippy::semicolon-if-nothing-returned` implied by `-D clippy::pedantic`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned
```

Clippy currently wants the following instead, which is unnecessary and more verbose as formatted by rustfmt.

```rust
let _d = || unsafe {
    ptr::drop_in_place(s.as_mut_ptr());
};
```

FYI @1c3t3a @mbartlett21 

### Meta

- `cargo clippy -V`: clippy 0.1.54 (c79419a 2021-06-04)
- `rustc -Vv`:
  ```
  rustc 1.54.0-nightly (c79419af0 2021-06-04)
  binary: rustc
  commit-hash: c79419af0721c614d050f09b95f076da09d37b0d
  commit-date: 2021-06-04
  host: x86_64-unknown-linux-gnu
  release: 1.54.0-nightly
  LLVM version: 12.0.1
  ```

### PR DESCRIPTION
**Pull request by 1c3t3a**
Currently the [`semicolon_if_nothing_returned`](https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned) lint fires in unwanted situations where a block only spans one line. An example of this was given in #7324. This code:

```rust
use std::mem::MaybeUninit;
use std::ptr;

fn main() {
    let mut s = MaybeUninit::<String>::uninit();
    let _d = || unsafe { ptr::drop_in_place(s.as_mut_ptr()) };
}
```

yields the following clippy error:
```
error: consider adding a `;` to the last statement for consistent formatting
 --> src/main.rs:6:26
  |
6 |     let _d = || unsafe { ptr::drop_in_place(s.as_mut_ptr()) };
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `ptr::drop_in_place(s.as_mut_ptr());`
  |
  = note: `-D clippy::semicolon-if-nothing-returned` implied by `-D clippy::pedantic`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned
```

I updated the lint to check if the statement is inside an `unsafe` block, a closure or a normal block and if the block only spans one line, in that case the lint is not emitted.

This closes #7324. 

changelog: enhanced semicolon if nothing returned according to  #7324.

### CODE REVIEW

**Comment by flip1995** [comment]
Please rebase on master. You have a commit in this PR that is already on master.

**Comment by 1c3t3a** [comment]
Sorry, just rebased moments before the PR, guess it was bad timing...

**Comment by 1c3t3a** [comment]
Looking at #7322, this PR would favor the `unsafe { f(x) };` style. But as described in the issue, it might be better to have this in a separate, configurable lint.

**Comment by 1c3t3a** [comment]
Alright, I added the changes discussed on Zulip, this PR is now ready for review @flip1995!

File: clippy_lints/src/semicolon_if_nothing_returned.rs [hunk 1]
```diff
@@ -66,3 +69,24 @@ impl LateLintPass<'_> for SemicolonIfNothingReturned {
         }
     }
 }
+
+/// Check if this block is inside a closure or an unsafe block or a normal on the same line.
+fn check_if_inside_block_on_same_line<'tcx>(
```
[the remark this answers is not in the document]
Comment (1c3t3a) [reply clippy_lints/src/semicolon_if_nothing_returned.rs:74]: Mhh... you're right, I tested this and thought about it but this makes sense as I only have such situations when the block is on a single line. I have thought too complicated! I will change the code and remove the function.

**Review by flip1995 — APPROVED** [review]
Thanks!

**Comment by 1c3t3a** [comment]
> Looking at #7322, this PR would favor the `unsafe { f(x) };` style. But as described in the issue, it might be better to have this in a separate, configurable lint.

What do you think about that @flip1995? Same lint or a new one?

**Comment by flip1995** [comment]
That should be a new pedantic lint.

**Pull request merged**